### PR TITLE
Create multi-tenant DI sample

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Connects to https://github.com/Particular/DeveloperExperience/issues/37

 * Create a new sample showing DI in the multi-tenant app
 * Simplify the pipeline UoW sample by removing concepts related to multi-tenancy

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
I discovered that this sample was actually used in a number of places for its UoW content so I'll revert the change and bring it back. I'll just remove all references to multi-tenancy.

**Comment by SzymonPobiega** [comment]
@mauroservienti this is ready for review.

File: samples/multi-tenant/di/sample.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+---
+title: Injecting tenant-aware components to the message handlers
+summary: Shows how to configure IoC to inject tenant-aware components to the message handlers
```
Comment (mauroservienti) [inline samples/multi-tenant/di/sample.md]:
```suggestion
summary: Shows how to configure IoC to inject tenant-aware components to message handlers
```

File: samples/multi-tenant/di/sample.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
+---
+title: Injecting tenant-aware components to the message handlers
```
Comment (mauroservienti) [inline samples/multi-tenant/di/sample.md]:
```suggestion
title: Injecting tenant-aware components to message handlers
```

File: samples/multi-tenant/di/sample.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+summary: Shows how to configure IoC to inject tenant-aware components to the message handlers
+reviewed: 2019-02-07
+component: Core
+---
+
+
+## Introduction
+
+This sample shows how to configure the dependency injection infrastructure built into NServiceBus to instantiate and inject tenant-aware components. Such approach allows the code in the handler to focus on the actual business logic without having to deal with multi-tenant aspect of the problem.
```
Comment (mauroservienti) [inline samples/multi-tenant/di/sample.md]:
```suggestion
This sample shows how to configure the dependency injection infrastructure built-in into NServiceBus to instantiate and inject tenant-aware components. Such approach allows code in handlers to focus on the actual business logic without having to deal with multi-tenant aspects of the problem.
```

File: samples/multi-tenant/di/sample.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
+---
+title: Injecting tenant-aware components to the message handlers
+summary: Shows how to configure IoC to inject tenant-aware components to the message handlers
+reviewed: 2019-02-07
+component: Core
```
Comment (mauroservienti) [inline samples/multi-tenant/di/sample.md]: A link to behaviors documentation as related would be a nice to have

File: samples/multi-tenant/di/sample.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+The following code is needed to register the behavior in the receive pipeline.
+
+snippet: configuration
+
+NOTE: The lifecycle has to be specified as `DependencyLifecycle.InstancePerUnitOfWork` in order to ensure there is a single instance of session created for each message being handled.
+
+
+### Message handlers
+
+One of the benefits of a unit of work is that multiple handlers for the same message will share the same session and commit/rollback together. This is how the handlers look:
```
Comment (mauroservienti) [inline samples/multi-tenant/di/sample.md]:
```suggestion
One of the benefits of a unit of work is that multiple handlers for the same message will share the same session and commit/rollback together. This is how handlers look like:
```

**Comment by mauroservienti** [comment]
Once this is merged we'll end up with the following samples:

- a basic pipeline behavior sample
- a tenant aware DI based sample

What about adding a tenant aware sample based on pipeline and context bag? (as the one we actually have)

**Comment by SzymonPobiega** [comment]
@mauroservienti I think that would be an overkill. The people I talked to during the multi-tenant research TF expected a way to just inject proper repository object into their handlers. They prefer this way because it is easier to test than the one that involves pipeline context.

File: samples/multi-tenant/di/sample.md (hunk 3 above)
Comment (SzymonPobiega) [reply samples/multi-tenant/di/sample.md]: I made it plural

**Comment by SzymonPobiega** [comment]
@mauroservienti I applied your suggestions.

**Pull request merged**